presto-main/src/main/java/com/facebook/presto/cost/SpatialJoinStatsRule.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.SpatialJoinNode;
+
+import java.util.Optional;
+
+import static com.facebook.presto.cost.PlanNodeStatsEstimate.UNKNOWN_STATS;
+import static com.facebook.presto.sql.planner.plan.Patterns.spatialJoin;
+import static java.util.Objects.requireNonNull;
+
+public class SpatialJoinStatsRule
+        extends SimpleStatsRule<SpatialJoinNode>
+{
+    private static final Pattern<SpatialJoinNode> PATTERN = spatialJoin();
+
+    private final FilterStatsCalculator filterStatsCalculator;
+
+    public SpatialJoinStatsRule(FilterStatsCalculator filterStatsCalculator, StatsNormalizer normalizer)
+    {
+        super(normalizer);
+        this.filterStatsCalculator = requireNonNull(filterStatsCalculator, "filterStatsCalculator is null");
+    }
+
+    @Override
+    protected Optional<PlanNodeStatsEstimate> doCalculate(SpatialJoinNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate leftStats = sourceStats.getStats(node.getLeft());
+        PlanNodeStatsEstimate rightStats = sourceStats.getStats(node.getRight());
+        PlanNodeStatsEstimate crossJoinStats = crossJoinStats(node, leftStats, rightStats);
+
+        switch (node.getType()) {
+            case INNER:
+                return Optional.of(filterStatsCalculator.filterStats(crossJoinStats, node.getFilter(), session, types));
```

Comment:
I'm working on removing the code that makes a wild `0.9` wild guess if nothing is know about the filters. But for now it will definitely produce something very wild, like `cross_product * 0.9`. But anyhow that is a exiting behaviour. Les't keep this class as is, so we can make sure that once we support spatial filter in filter estimates it will be enabled by default.